File: 86-partitionList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @param {integer} x
    # @return {ListNode}
    def partition(self, head, x):
        if head is None or head.next is None: return head
        small_start = ListNode(-1); large_start = ListNode(-1) 
        small_start.next = large_start.next = head
        stmp = small_start; ltmp = large_start
        while stmp.next:
              if stmp.next.val < x:
                 stmp = stmp.next
                 ltmp.next = ltmp.next.next
              else:
                 ltmp = ltmp.next
                 stmp.next = stmp.next.next
        stmp.next = large_start.next
        return small_start.next

    def partition2(self, head, x):
        head1 = ListNode(0)
        head2 = ListNode(0)
        Tmp = head
        phead1 = head1
        phead2 = head2
        while Tmp:
            if Tmp.val < x:
                phead1.next = Tmp
                Tmp = Tmp.next
                phead1 = phead1.next
                phead1.next = None
            else:
                phead2.next = Tmp
                Tmp = Tmp.next
                phead2 = phead2.next
                phead2.next = None
        phead1.next = head2.next
        head = head1.next
        return head


def print_llst(head):
    llst = ''
    while head:
          llst += str(head.val)
          if head.next:
             llst += '->'
          head = head.next
    print llst

if __name__ == '__main__':
   s = Solution()
   test = ListNode(1)
   test.next = ListNode(4)
   test.next.next = ListNode(3)
   test.next.next.next = ListNode(2)
   test.next.next.next.next = ListNode(5)
   test.next.next.next.next.next = ListNode(2)
   print_llst(s.partition(test, 3))

File: C++/86-partitionList.cpp
#include "include/common.h"
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        Utils u;
        ListNode *dummy = new ListNode(-1), *larger = new ListNode(0), *largeStart;
        dummy->next = head;
        head        = dummy; // notice here!, dont use dummy->next
        largeStart  = larger;  // notice here!, dont use larger->next
        while (dummy->next != nullptr) {
            if (dummy->next->val < x) {
                dummy = dummy->next;
            } else {
                ListNode *next = dummy->next->next;
                dummy->next->next = nullptr;
                larger->next = dummy->next;
                dummy->next = next;
                larger = larger->next;
            };
        };
        dummy->next = largeStart->next;
        return head->next;
    }
};

int main(void) {
    Solution s;
    Utils u;
    ListNode l1(1);
    ListNode l2(4);
    ListNode l3(3);
    ListNode l4(2);
    ListNode l5(5);
    ListNode l6(2);
    (&l1)->next = (&l2);
    (&l2)->next = (&l3);
    (&l3)->next = (&l4);
    (&l4)->next = (&l5);
    (&l5)->next = (&l6);
    u.printLinkedList(s.partition(&l1, 3));
    return 0;
}